File: client-admin/types/coupon.ts
import { CategoryForSelect } from "./category";

// coupon.types.ts
export interface Coupon {
  _id: string;
  code: string;
  name: string;
  description: string;
  type: CouponType; // chỉ còn 'percentage' | 'fixed'
  value: number;
  minOrderValue: number;
  maxDiscountValue: number;
  applyTo: CouponApplyTo; // chỉ còn 'order'
  applyToAllCategories: true;
  applicableProducts: [];
  applicableCategories: [];
  // Thời gian
  startDate: string;
  endDate: string;
  // Giới hạn sử dụng
  maxUsageCount: number;
  usageCount: number;
  maxUsagePerUser: number;
  // Trạng thái & flags
  isActive: boolean;
  status: CouponStatus;
  priority: CouponPriority;
  userCondition: UserCondition;
  stackable: boolean;
  createdAt: string;
  updatedAt: string;
}

// type cho type và status
export type CouponType = "percentage" | "fixed"

export type CouponStatus = "active" | "scheduled" | "expired" | "exhausted" | "inactive"

export type UserCondition = "all" | "new" | "vip"

export type CouponPriority = "low" | "medium" | "high"

export type CouponApplyTo = "order"


export type CouponAvailable = {
  _id: string
  code: string
  name: string
  description: string
  type: CouponType
  value: number
  maxDiscountValue: number
  minOrderValue: number
  usageCount: number
  maxUsageCount: number
  maxUsagePerUser: number
  stackable: boolean
  eligible: boolean
  reason: string
  requireLoginToUse: boolean
  estimatedDiscount: number
}

File: server/app/models/coupon.py
from app.utils.format_time import to_utc_iso

def coupon_valueable(coupon):
    return {
        "_id": coupon["_id"],
        "code": coupon["code"],
        "name": coupon["name"],
        "description": coupon["description"],
        "type": coupon["type"],
        "value": coupon["value"],
        "maxDiscountValue": coupon["maxDiscountValue"],
        "minOrderValue": coupon["minOrderValue"],
        "usageCount": coupon["usageCount"],
        "maxUsageCount": coupon["maxUsageCount"],
        "maxUsagePerUser": coupon["maxUsagePerUser"],
        "stackable": coupon["stackable"],
        "eligible": coupon.get("eligible", True),
        "reason": coupon.get("reason", ""),
        "requireLoginToUse": coupon.get("requireLoginToUse", False),
        "estimatedDiscount": coupon.get("estimatedDiscount", 0),
    }

